tests/logging/test_remote_handler.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from twisted.test.proto_helpers import AccumulatingProtocol
+
+from synapse.logging import RemoteHandler
+
+from tests.server import FakeTransport
+from tests.unittest import DEBUG, HomeserverTestCase
+
+
+class StructuredLoggingTestBase:
+    """
+    Test base that registers a cleanup handler to reset the stdlib log handler
+    to 'unset'.
+    """
+
+    def prepare(self, reactor, clock, hs):
+        def _cleanup():
+            logging.getLogger("synapse").setLevel(logging.NOTSET)
+
+        self.addCleanup(_cleanup)
+
+
+def connect_logging_client(reactor, client_id):
+    # This is essentially tests.server.connect_client, but disabling autoflush on
+    # the client transport. This is necessary to avoid an infinite loop due to
+    # sending of data via the logging transport causing additional logs to be
+    # written.
+    factory = reactor.tcpClients.pop(client_id)[2]
+    client = factory.buildProtocol(None)
+    server = AccumulatingProtocol()
+    server.makeConnection(FakeTransport(client, reactor))
+    client.makeConnection(FakeTransport(server, reactor, autoflush=False))
+
+    return client, server
+
+
+class RemoteHandlerTestCase(StructuredLoggingTestBase, HomeserverTestCase):
+    @DEBUG
```

Comment:
Are these necessary for the tests? If so I'd say we should use another method as this looks like someone has accidentally committed @DEBUG

Reply:
I believe they are, but I'll check!

Reply:
They *are* required, although we can change them to `@INFO` if that's more palatable? Otherwise I can look deeper into what's going on, but I suspect something is poking at the logging infrastructure in tests to disable logging.

Reply:
Yeah, by default the log levels are set at `ERROR`. My main objection is that `@DEBUG` and co are mostly helper functions that let you easily turn on debug logs while testing something, and as such usages generally never should be committed.

I think you can manually set `loglevel = logging.DEBUG` as a class attribute to turn on debug logging for all tests in the class, which would also allow a comment to say why its necessary

Reply:
Actually, we just need to set the log level of the logger we manually create. See 650bb09f88353e8513adb6b91cb5492663167c96. I think this is much clearer.